### PR TITLE
Compile regex on demand not in init

### PR DESCRIPTION
**Pull request by rhatdan**
Every app is paying the price for these compilations on startup so make only the users of the functions pay the price.

ValidateDNSSearch is not used anywhere so remove it.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@alexlarsson @vrothberg PTAL

**Comment by alexlarsson** [comment]
Maybe we can use https://github.com/charlievieth/reonce instead?

**Comment by alexlarsson** [comment]
Using this it would be less work to go a wider replacement of all regexp.MustCompile global vars.

File: internal/opts/opts.go [hunk 1]
```diff
@@ -222,26 +216,6 @@ func ValidateIPAddress(val string) (string, error) {
 	return "", fmt.Errorf("%s is not an ip address", val)
 }
 
-// ValidateDNSSearch validates domain for resolvconf search configuration.
```
Comment (vrothberg) [inline internal/opts/opts.go:225]: Why remove the code?  Please put it in a different commit in case the code is dead.

File: pkg/stringid/stringid.go [hunk 2]
```diff
@@ -19,15 +19,25 @@ import (
...
 
 var (
-	validShortID = regexp.MustCompile("^[a-f0-9]{12}$")
-	validHex     = regexp.MustCompile(`^[a-f0-9]{64}$`)
+	validShortID *regexp.Regexp
+	validHex     *regexp.Regexp
 
 	rngLock sync.Mutex
 	rng     *rand.Rand // A RNG with seeding properties we control. It can only be accessed with randLock held.
+	once    sync.Once
```
Comment (vrothberg) [inline pkg/stringid/stringid.go:27]:
```suggestion
	onceRegex    sync.Once
```

Just to add some context to the variable.

**Comment by vrothberg** [comment]
> Maybe we can use https://github.com/charlievieth/reonce instead?

It doesn't look maintained.

**Comment by rhatdan** [comment]
I like the idea of reonce, and would love to see that get into the mainstream package, but I am afraid to add an unmaintained package.

@alexlarsson @vrothberg @giuseppe @flouthoc @nalind @umohnani8 PTAL

**Review by ashley-cui — COMMENTED** [review]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by alexlarsson** [comment]
While i agree in theory that "unmaintained" packages are bad, i'm not sure that package is actuallyt unmaintained. It got a commit in august 2022, for example. And fundamentally, the package is extremely simple, all it does is forward some calls wrapped around a call to sync.Once. I don't see why this would really need to change much.

Honestly, reimplementing reonce inside containers/common is probably going to end up being less code than open-coding all the regexes in the codebase in a similar way to this merge request.
